apps/dashboard/src/utils/schema.ts

```diff
@@ -0,0 +1,167 @@
+import * as z from 'zod';
+import { WorkflowTestDataResponseDto } from '@novu/shared';
+import { capitalize } from './string';
+
+type JSONSchema = WorkflowTestDataResponseDto['to'];
+
+type ZodValue =
+  | z.AnyZodObject
+  | z.ZodString
+  | z.ZodNumber
+  | z.ZodEffects<z.ZodTypeAny>
+  | z.ZodDefault<z.ZodTypeAny>
+  | z.ZodEnum<[string, ...string[]]>
+  | z.ZodOptional<z.ZodTypeAny>;
+
+const handleStringFormat = ({ value, key, format }: { value: z.ZodString; key: string; format: string }) => {
+  if (format === 'email') {
+    return value.email(`${capitalize(key)} must be a valid email`);
+  } else if (format === 'uri') {
+    return value
+      .transform((val) => (val === '' ? undefined : val))
+      .refine((val) => !val || z.string().url().safeParse(val).success, {
+        message: `${capitalize(key)} must be a valid URI`,
+      });
+  }
+
+  return value;
+};
+
+const handleStringPattern = ({ value, key, pattern }: { value: z.ZodString; key: string; pattern: string }) => {
+  return value
+    .transform((val) => (val === '' ? undefined : val))
+    .refine((val) => !val || z.string().regex(new RegExp(pattern)).safeParse(val).success, {
+      message: `${capitalize(key)} must be a valid value`,
+    });
+};
+
+const handleStringEnum = ({ key, enumValues }: { key: string; enumValues: [string, ...string[]] }) => {
+  return z.enum(enumValues, { message: `${capitalize(key)} must be one of ${enumValues.join(', ')}` });
+};
+
+const handleStringType = ({
+  key,
+  format,
+  pattern,
+  enumValues,
+  defaultValue,
+  requiredFields,
+}: {
+  key: string;
+  format?: string;
+  pattern?: string;
+  enumValues?: unknown;
+  defaultValue?: unknown;
+  requiredFields: Readonly<Array<string>>;
+}) => {
+  const isRequired = requiredFields.includes(key);
+
+  let stringValue:
+    | z.ZodString
+    | z.ZodEffects<z.ZodTypeAny>
+    | z.ZodEnum<[string, ...string[]]>
+    | z.ZodDefault<z.ZodTypeAny> = z.string();
+
+  if (format) {
+    stringValue = handleStringFormat({
+      value: stringValue,
+      key,
+      format,
+    });
+  } else if (pattern) {
+    stringValue = handleStringPattern({
+      value: stringValue,
+      key,
+      pattern,
+    });
+  } else if (enumValues) {
+    stringValue = handleStringEnum({
+      key,
+      enumValues: enumValues as [string, ...string[]],
+    });
+  } else if (isRequired) {
+    stringValue = stringValue.min(1, `${capitalize(key)} is required`);
+  }
+
+  if (defaultValue) {
+    stringValue = stringValue.default(defaultValue as string);
+  }
+
+  // remove empty strings
+  return stringValue.transform((val) => (val === '' ? undefined : val));
+};
+
+/**
+ * Transform JSONSchema to Zod schema.
+ * The function will recursively build the schema based on the JSONSchema object.
+ * It removes empty strings and objects with empty required fields during the transformation phase after parsing.
+ */
+export const buildDynamicZodSchema = (obj: JSONSchema): z.AnyZodObject => {
```

Comment:
This is the util used to parse the `dataSchema` (which is JSONSchema type) to `Zod.` The converted schema is used in the form for the validation and then transformation after the parsing (success validation) phase. The transformed value is returned in the `onSubmit` handler.

Reply:
if you need zod why do we serve jsonSchema? @SokratisVidros

Reply:
I guess that the FE prefers to work with Zod for client-side validations so they transform the JSON schema to zod.

Reply:
@tatarco, the Zod schema is not serializable; please check [this discussion](https://github.com/colinhacks/zod/discussions/2030). There are tools that can convert Zod <> JSON schema and vice versa, but they are not "perfect" and require some runtime evaluation.

Reply:
@tatarco is right. Why didn't we use the [JSON Schema binding](https://github.com/vtex/react-hook-form-jsonschema) for React-hook form?